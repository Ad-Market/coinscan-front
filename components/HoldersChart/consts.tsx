export enum HodlersChartGroupings {
  BY_DAY,
  BY_WEEK,
  BY_MONTH,
}

export const LINE_CHART_GROUPS = [HodlersChartGroupings.BY_DAY];
export const BAR_CHART_GROUPS = [
  HodlersChartGroupings.BY_WEEK,
  HodlersChartGroupings.BY_MONTH,
];
