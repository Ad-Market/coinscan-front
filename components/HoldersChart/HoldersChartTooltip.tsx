import { TooltipProps } from 'recharts';
import { endOfWeek, fromUnixTime, format, parse } from 'date-fns';
import { XAxisProps } from 'recharts';
import {
  ValueType,
  NameType,
  Props as DefaultTooltipContentProps,
} from 'recharts/types/component/DefaultTooltipContent';
import { HodlersChartGroupings } from './consts';

type XAxisTickFormatterFn = XAxisProps['tickFormatter'];

type XAxisTickFormatters = {
  [key in HodlersChartGroupings]: XAxisTickFormatterFn;
};

const getLocaleDateStringFromUnixTime = (value: string) =>
  fromUnixTime(parseInt(value, 10)).toLocaleDateString();

export const HODLERS_CHART_XAXIS_TICK_FORMATTERS: XAxisTickFormatters = {
  [HodlersChartGroupings.BY_DAY]: (value) =>
    getLocaleDateStringFromUnixTime(value),
  [HodlersChartGroupings.BY_WEEK]: (value) =>
    getLocaleDateStringFromUnixTime(value),
  [HodlersChartGroupings.BY_MONTH]: (value) => {
    return format(fromUnixTime(parseInt(value, 10)), 'MM.y');
  },
};

type TooltipLabelFormatterFn = DefaultTooltipContentProps<
  ValueType,
  NameType
>['labelFormatter'];

type TooltipLabelFormatters = {
  [key in HodlersChartGroupings]: TooltipLabelFormatterFn;
};

export const HODLERS_CHART_TOOLTIP_LABEL_FORMATTERS: TooltipLabelFormatters = {
  [HodlersChartGroupings.BY_DAY]: (label) =>
    getLocaleDateStringFromUnixTime(label),
  [HodlersChartGroupings.BY_WEEK]: (_, payload) => {
    const dataDatetime = fromUnixTime(parseInt(payload[0].payload.name, 10));
    const start = dataDatetime.toLocaleDateString();
    const end = endOfWeek(dataDatetime, {
      weekStartsOn: 1,
    }).toLocaleDateString();
    return `${start} - ${end}`;
  },
  [HodlersChartGroupings.BY_MONTH]: (_, payload) => {
    const dataDatetime = fromUnixTime(parseInt(payload[0].payload.name, 10));
    const formattedDate = format(dataDatetime, 'MM.y');
    return `${formattedDate}`;
  },
};

const HoldersChartTooltip = ({
  active,
  payload,
  label,
  labelFormatter,
}: TooltipProps<ValueType, NameType>) => {
  if (active && payload && payload.length > 0) {
    return (
      <div>
        {labelFormatter ? labelFormatter(label, payload) : label}:{' '}
        {payload[0].value}
      </div>
    );
  }

  return null;
};

export default HoldersChartTooltip;
