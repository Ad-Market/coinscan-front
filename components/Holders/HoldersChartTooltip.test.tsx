import { render, screen } from '@testing-library/react';

import HoldersChartTooltip, {
  HODLERS_CHART_TOOLTIP_LABEL_FORMATTERS,
} from './HoldersChartTooltip';
import { HodlersChartGroupings } from './consts';

test('HoldersTooltip renders correctly for daily grouping', () => {
  const payload = [
    {
      payload: { count: 6293, name: '1638144000' },
    },
  ];

  const { container } = render(
    <HoldersChartTooltip
      active
      label={1654473600}
      payload={payload}
      labelFormatter={
        HODLERS_CHART_TOOLTIP_LABEL_FORMATTERS[HodlersChartGroupings.BY_DAY]
      }
    />
  );

  expect(container).toMatchSnapshot('rendered');
});

test('HoldersTooltip renders correctly for weekly grouping', () => {
  const payload = [
    {
      payload: { count: 5393, name: '1638144000' },
    },
  ];

  const { container } = render(
    <HoldersChartTooltip
      active
      label={payload[0].payload.name}
      payload={payload}
      labelFormatter={
        HODLERS_CHART_TOOLTIP_LABEL_FORMATTERS[HodlersChartGroupings.BY_WEEK]
      }
    />
  );

  expect(container).toMatchSnapshot('rendered');
});

test('HoldersTooltip renders correctly for monthly grouping', () => {
  const payload = [
    {
      payload: {
        count: 5273,
        name: '1635724800',
      },
    },
  ];

  const { container } = render(
    <HoldersChartTooltip
      active
      label={payload[0].payload.name}
      payload={payload}
      labelFormatter={
        HODLERS_CHART_TOOLTIP_LABEL_FORMATTERS[HodlersChartGroupings.BY_MONTH]
      }
    />
  );

  expect(container).toMatchSnapshot('rendered');
});
