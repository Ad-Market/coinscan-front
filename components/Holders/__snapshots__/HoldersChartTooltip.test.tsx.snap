// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoldersTooltip renders correctly for daily grouping: rendered 1`] = `
<div>
  <div
    class="MuiBox-root css-r6fq3j"
  >
    6/6/2022
    :
     
    6293
  </div>
</div>
`;

exports[`HoldersTooltip renders correctly for monthly grouping: rendered 1`] = `
<div>
  <div
    class="MuiBox-root css-r6fq3j"
  >
    11.2021
    :
     
    5273
  </div>
</div>
`;

exports[`HoldersTooltip renders correctly for weekly grouping: rendered 1`] = `
<div>
  <div
    class="MuiBox-root css-r6fq3j"
  >
    11/29/2021 - 12/5/2021
    :
     
    5393
  </div>
</div>
`;
